neqo-transport/src/path.rs

```diff
@@ -106,4 +535,197 @@ impl Path {
+            builder.encode_varint(FRAME_TYPE_PATH_CHALLENGE);
+            builder.encode(&data);
+
+            // As above, no recovery token.
+            stats.path_challenge += 1;
+            stats.all += 1;
+
+            self.state = ProbeState::Probing {
+                probe_count,
+                data,
+                mtu,
+                sent: now,
+            };
+            true
+        } else {
+            resp_sent
+        }
+    }
+
+    /// Process a timer for this path.
+    /// This returns true if the path is viable and can be kept alive.
+    pub fn process_timeout(&mut self, now: Instant, pto: Duration) -> bool {
+        if let ProbeState::Probing { sent, .. } = &self.state {
+            if now >= *sent + pto {
+                self.probe();
+            }
+        }
+        if let ProbeState::Failed = self.state {
+            // Retire failed paths immediately.
+            false
+        } else if self.primary {
+            // Keep valid primary paths otherwise.
+            true
+        } else if let ProbeState::Valid = self.state {
+            // Retire validated, non-primary paths.
+            // Allow more than `MAX_PATH_PROBES` times the PTO so that an old
+            // path remains around until after a previous path fails.
+            let count = u32::try_from(MAX_PATH_PROBES + 1).unwrap();
+            self.validated.unwrap() + (pto * count) > now
+        } else {
+            // Keep paths that are being actively probed.
+            true
+        }
+    }
+
+    /// Return the next time that this path needs servicing.
+    /// This only considers retransmissions of probes, not cleanup of the path.
+    /// If there is no other activity, then there is no real need to schedule a
+    /// timer to cleanup old paths.
+    pub fn next_timeout(&self, pto: Duration) -> Option<Instant> {
+        if let ProbeState::Probing { sent, .. } = &self.state {
+            Some(*sent + pto)
+        } else {
+            None
+        }
+    }
+
+    /// Record received bytes for the path.
+    pub fn add_received(&mut self, count: usize) {
+        self.received_bytes = self.received_bytes.saturating_add(count);
+    }
+
+    /// Record sent bytes for the path.
+    pub fn add_sent(&mut self, count: usize) {
+        self.sent_bytes = self.sent_bytes.saturating_add(count);
+    }
+
+    /// Get the number of bytes that can be written to this path.
+    pub fn amplification_limit(&self) -> usize {
+        if matches!(self.state, ProbeState::Failed) {
+            0
+        } else if self.is_valid() {
+            usize::MAX
+        } else if let Some(limit) = self.received_bytes.checked_mul(3) {
+            let budget = if limit == 0 {
+                // If we have received absolutely nothing thus far, then this endpoint
+                // is the one initiating communication on this path.  Allow enough space for probing.
+                self.mtu() * 5
```

Comment:
There's a gotcha here: we're going to probe any path we migrate to and send a padded datagram.  We should get 1200 bytes or more back.  If we don't, we're going to be bound by the amplification limit as a client.  This isn't ideal, but I think that it's tolerable.  We will ignore this amplification limit for any PTO that we send, so we aren't completely deadlocked.  It's a little annoying, that's all.

Reply:
ok.